src/THORChain/Swap.cpp

```diff
@@ -237,22 +202,66 @@ std::pair<int, std::string> Swap::buildBinance([[maybe_unused]] Chain toChain, [
     token.set_amount(amount);
     {
         Binance::Address fromAddressBin;
-        Binance::Address::decode(fromAddress, fromAddressBin);
+        Binance::Address::decode(mFromAddress, fromAddressBin);
         auto input_ = order.add_inputs();
         input_->set_address(fromAddressBin.getKeyHash().data(), fromAddressBin.getKeyHash().size());
         *input_->add_coins() = token;
     }
     {
         Binance::Address vaultAddressBin;
-        Binance::Address::decode(vaultAddress, vaultAddressBin);
+        Binance::Address::decode(mVaultAddress, vaultAddressBin);
         auto output = order.add_outputs();
         output->set_address(vaultAddressBin.getKeyHash().data(), vaultAddressBin.getKeyHash().size());
         *output->add_coins() = token;
     }
 
     auto serialized = input.SerializeAsString();
     out.insert(out.end(), serialized.begin(), serialized.end());
-    return std::make_pair<int, std::string>(0, "");
+    return std::make_tuple(std::move(out), 0, "");
 }
 
+std::tuple<Data, int, std::string> SwapBuilder::buildEth(uint64_t amount, const std::string& memo) {
+    Data out;
+    auto input = Ethereum::Proto::SigningInput();
+    const auto& toTokenId = mToAsset.token_id();
+    // some sanity check / address conversion
+    Data vaultAddressBin = ethAddressStringToData(mVaultAddress);
+    if (!Ethereum::Address::isValid(mVaultAddress) || vaultAddressBin.size() != Ethereum::Address::size) {
+        return std::make_tuple(Data(), static_cast<int>(Proto::ErrorCode::Error_Invalid_vault_address), "Invalid vault address: " + mVaultAddress);
+    }
+    if (!Ethereum::Address::isValid(*mRouterAddress)) {
+        return std::make_tuple(Data(), static_cast<int>(Proto::ErrorCode::Error_Invalid_router_address), "Invalid router address: " + *mRouterAddress);
+    }
+    Data toAssetAddressBin = ethAddressStringToData(toTokenId);
+
+    // Following fields must be set afterwards, before signing ...
+    const auto chainId = store(uint256_t(0));
+    input.set_chain_id(chainId.data(), chainId.size());
+    const auto nonce = store(uint256_t(0));
+    input.set_nonce(nonce.data(), nonce.size());
+    const auto gasPrice = store(uint256_t(0));
+    input.set_gas_price(gasPrice.data(), gasPrice.size());
+    const auto gasLimit = store(uint256_t(0));
+    input.set_gas_limit(gasLimit.data(), gasLimit.size());
+    input.set_private_key("");
+    // ... end
+
+    input.set_to_address(*mRouterAddress);
+    auto& transfer = *input.mutable_transaction()->mutable_contract_generic();
+    auto func = Ethereum::ABI::Function("deposit", std::vector<std::shared_ptr<Ethereum::ABI::ParamBase>>{
+                                                       std::make_shared<Ethereum::ABI::ParamAddress>(vaultAddressBin),
+                                                       std::make_shared<Ethereum::ABI::ParamAddress>(toAssetAddressBin),
```

Comment:
I think this was wrong here in the original, I think this is supposed to be fromAssetAddress, but it has to be verified (so far this was never really used).

Reply:
> I think this was wrong here in the original, I think this is supposed to be fromAssetAddress, but it has to be verified (so far this was never really used).

should we address this during erc20 token addition?

Reply:
@Milerius @catenocrypt should we fix this? others LGTM

Reply:
Need more time to clarify/test; added note to #2744, deferring

Reply:
Need more time to clarify/test; added note to #2744, deferring